core/src/main/groovy/com/muwire/core/files/PersisterFolderService.groovy

```diff
@@ -0,0 +1,141 @@
+import java.util.concurrent.Executors
+import java.util.concurrent.ThreadFactory
+import java.util.logging.Level
+
+/**
+ * A persister that stores information about the files shared using
+ * individual JSON files in directories.
+ *
+ * The absolute path's 32bit hash to the shared file is used
+ * to build the directory and filename.
+ *
+ * This persister only starts working once the old persister has finished loading
+ * @see PersisterFolderService#getJsonPath
+ */
+@Log
+class PersisterFolderService extends BasePersisterService {
+
+    final static int CUT_LENGTH = 6
+
+    final File location
+    final EventBus listener
+    final int interval
+    final Timer timer
+    final ExecutorService persisterExecutor = Executors.newSingleThreadExecutor({ r ->
+        new Thread(r, "file persister")
+    } as ThreadFactory)
+
+    PersisterFolderService(File location, EventBus listener) {
+        this.location = location
+        this.listener = listener
+        this.interval = interval
+        timer = new Timer("file-folder persister timer", true)
+    }
+
+    void stop() {
+        timer.cancel()
+        persisterExecutor.shutdown()
+    }
+
+    void onPersisterDoneEvent(PersisterDoneEvent persisterDoneEvent) {
+        log.info("Old persister done")
+        load()
+    }
+
+    void onFileHashedEvent(FileHashedEvent hashedEvent) {
+        persistFile(hashedEvent.sharedFile)
+    }
+    /**
+     * Get rid of the json of unshared files
+     * @param unsharedEvent
+     */
+    void onFileUnsharedEvent(FileUnsharedEvent unsharedEvent) {
+        def jsonPath = getJsonPath(unsharedEvent.unsharedFile)
+        def jsonFile = jsonPath.toFile()
+        if(jsonFile.isFile()){
+            jsonFile.delete()
+        }
+    }
+    void onFileLoadedEvent(FileLoadedEvent loadedEvent) {
+        if(loadedEvent.source == "PersisterService"){
+            log.info("Migrating persisted file from PersisterService: "
+                    + loadedEvent.loadedFile.file.absolutePath.toString())
+            persistFile(loadedEvent.loadedFile)
+        }
+    }
+
+    void load() {
+        log.fine("Loading...")
+        Thread.currentThread().setPriority(Thread.MIN_PRIORITY)
+
+        if (location.exists() && location.isDirectory()) {
+            try {
+                _load()
+            }
+            catch (IllegalArgumentException e) {
+                log.log(Level.WARNING, "couldn't load files", e)
+            }
+        } else {
+            location.mkdirs()
+            listener.publish(new AllFilesLoadedEvent())
+        }
+        loaded = true
+    }
+
+    /**
+     * Loads every JSON into memory
+     */
+    private void _load() {
+        int loaded = 0
+        def slurper = new JsonSlurper()
+        Files.walk(location.toPath())
+                .filter({ it.fileName.endsWith(".json") })
+                .forEach({
+                    def parsed = slurper.parse it.toFile()
+                    def event = fromJson parsed
```

Comment:
Don't you need to change the `source` field in the event here?  Otherwise you're re-persisting the each `SharedFile` on load

Reply:
Ahh... good catch. It used to be `sourceClass` but for some reason `this.getClass()` and `this.class` kept returning `Class`. I'll fix that

Reply:
Done https://github.com/zlatinb/muwire/pull/37/commits/a2b37ef567655595019d08ab11936088693edbd8

Reply:
Done https://github.com/zlatinb/muwire/pull/37/commits/a2b37ef567655595019d08ab11936088693edbd8